File: src/priv/bajdcc/util/HashListMapEx2.java
package priv.bajdcc.util;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * 哈希表
 * 
 * @author bajdcc
 */
public class HashListMapEx2<K, V> {

	public HashMap<K, V> map = new HashMap<>();

	public ArrayList<K> list = new ArrayList<>();

	public boolean contains(K k) {
		return map.containsKey(k);
	}

	public void add(K k, V v) {
		V vv = map.put(k, v);
		if (vv == null) {
			list.add(k);
		}
	}

	public boolean put(K k, V v) {
		V vv = map.put(k, v);
		if (vv == null) {
			list.add(k);
			return false;
		} else {
			return true;
		}
	}

	public V get(K k) {
		return map.get(k);
	}

	public int indexOf(K k) {
		return list.indexOf(k);
	}
	
	public void pop() {
		int last = list.size() - 1;
		map.remove(list.get(last));
		list.remove(last);
	}
}

File: src/main/java/com/bajdcc/util/HashListMapEx2.kt
package com.bajdcc.util

/**
 * 双向哈希表
 * [K] -> [V]
 * @author bajdcc
 */
class HashListMapEx2<K, V> {

    private var map = mutableMapOf<K, V>()
    private var list = mutableListOf<K>()

    operator fun contains(k: K) = map.containsKey(k)

    fun add(k: K, v: V) {
        if (map.put(k, v) == null)
            list.add(k)
    }

    fun pop() {
        map.remove(list.get(list.size - 1))
        list.removeAt(list.size - 1)
    }

    operator fun get(k: K): V = map[k]!!

    fun copy() = HashListMapEx2<K, V>().also {
        it.map = this.map.toMutableMap()
        it.list = this.list.toMutableList()
    }

    fun toList() : List<V> = map.toList().map { it.second }
}